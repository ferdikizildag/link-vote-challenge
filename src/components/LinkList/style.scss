.orderSelect{
  margin-bottom: 10px;
}
