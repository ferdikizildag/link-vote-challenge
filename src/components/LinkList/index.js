import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import style from './style.module.scss';
import { VoteOrderEnum } from 'utils/enum';
import LinkItem from 'components/LinkItem';
import Pagination from 'components/Pagination';
import CustomModal from 'components/CustomModal';
import { toastr } from 'react-redux-toastr';
const pageSize = 5;

class LinkList extends Component {
    state = {
        visibility: false,
        orderByType: '',
        activePage: 0,
        modalData: {}
    }

    setModalVisibility = (value) => {
        this.setState({ visibility: value })
    }

    setModalData = (link) => {
        this.setState({ modalData: link })
    }

    setActivePage = (page) => {
        this.setState({ activePage: page })
    }

    setOrderByType = (type) => {
        this.setState({ orderByType: type })
    }

    hideModal = () => {
        this.setModalVisibility(false)
    }

    removeLink = () => {
        const { deleteLink } = this.props;
        const { modalData } = this.state;
        deleteLink(modalData)
        this.setModalVisibility(false)
        this.setModalData({})
        toastr.success('', modalData.name + ' Removed');
    }

    showModal = (visibility, item) => {
        this.setModalVisibility(visibility)
        this.setModalData(item)
    }

    orderBySelect = (e) => {
        this.setOrderByType(e.target.value)
        this.setActivePage(0)
    }

    sortByMostVoted = (x, y) => {
        const value = y.point - x.point;
        if (value === 0) {
            return this.sortByUpdatedAt(x, y);
        }
        return value;
    };

    sortByLessVoted = (x, y) => {
        const value = x.point - y.point;
        if (value === 0) {
            return this.sortByUpdatedAt(x, y);
        }
        return value;
    };

    sortByUpdatedAt = (x, y) => {
        return new Date(y.updatedAt).getTime() - new Date(x.updatedAt).getTime();
    };

    sortByCreatedAt = (x, y) => {
        return new Date(y.createdAt).getTime() - new Date(x.createdAt).getTime();
    };
    renderLinkItem = () => {
        const { data, updateVote } = this.props;
        const { orderByType, activePage } = this.state;
        if (data && data.length > 0) {
            let pageData = [];
            switch (orderByType) {
                case VoteOrderEnum.LESS_VOTED:
                    pageData = data.sort(this.sortByLessVoted);
                    break;
                case VoteOrderEnum.MOST_VOTED:
                    pageData = data.sort(this.sortByMostVoted);
                    break;
                default:
                    pageData = data.sort(this.sortByCreatedAt);
                    break;
            }

            pageData = pageData.slice(pageSize * activePage, pageSize * (activePage + 1));

            if (pageData.length === 0) {
                pageData = data.slice(pageSize * (activePage - 1), pageSize * activePage);
                this.setActivePage(activePage - 1);
            }
            return pageData.map((link, key) => (
                <LinkItem
                    key={key}
                    item={link}
                    removeClick={this.showModal}
                    updateVote={updateVote}
                />
            ));
        } else {
            return '';
        }
    };
    render() {
        const { data } = this.props;
        const { activePage, modalData, visibility } = this.state;
        const totalSize = Math.floor((data.length - 1) / pageSize);
        return (
            <div>
                <select className={style.orderSelect} onChange={this.orderBySelect}>
                    <option value="">Order by</option>
                    <option value={VoteOrderEnum.MOST_VOTED}>Most Voted (Z->A)</option>
                    <option value={VoteOrderEnum.LESS_VOTED}>Less Voted (A->Z)</option>
                </select>
                {this.renderLinkItem()}
                {totalSize >= 1 && (
                    <Pagination currentPage={activePage} totalSize={totalSize} setActivePage={this.setActivePage} />
                )}
                <CustomModal
                    question='Do you want to remove:'
                    header='Remove Link'
                    data={modalData}
                    onClickOk={this.removeLink}
                    onClickCancel={this.hideModal}
                    visibility={visibility}
                />
            </div>
        )
    }
}

LinkList.defaultProps = {
    data: []
};

LinkList.propTypes = {
    data: PropTypes.array,
    deleteLink: PropTypes.func,
    updateVote: PropTypes.func
};

export default LinkList;