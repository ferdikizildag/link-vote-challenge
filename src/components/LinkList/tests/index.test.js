import React from 'react';
import { mount } from 'enzyme';
import LinkList from 'components/LinkList';

describe('LinkList', () => {

    const deleteLinkMock = jest.fn();
    const updateVoteMock = jest.fn();
    const removeLinkMock = jest.fn();

    const linkMock = {
        id: '123123',
        point: 1,
        createdAt: new Date(),
        link: 'link',
        name: 'name',
        updatedAt: new Date(),
    };
    let linkListProps = {
        data: [],
        deleteLink: deleteLinkMock,
        updateVote: updateVoteMock,
        removeLink: removeLinkMock
    }

    it('renders without crashing', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();
        expect(instance.render()).toMatchSnapshot();
    });

    it('function setModalVisibility', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();
        instance.setModalVisibility(true);
        expect(instance.state.visibility).toEqual(true);
    });

    it('function setModalData', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();
        instance.setModalData(linkMock);
        expect(JSON.stringify(instance.state.modalData)).toEqual(JSON.stringify(linkMock));
    });

    it('sortByMostVoted of LinkList while points are not equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 1,
            updatedAt: dumUpdatedDate,
        };
        expect(instance.sortByMostVoted(dumLink1, dumLink2)).toEqual(1);
    });

    it('sortByMostVoted of LinkList while points are equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        instance.sortByUpdatedAt = jest.fn().mockReturnValue(1);
        await expect(instance.sortByMostVoted(dumLink1, dumLink2)).toEqual(1);
        expect(instance.sortByUpdatedAt).toHaveBeenCalledWith(dumLink1, dumLink2);
    });

    it('sortByLessVoted of LinkList while points are not equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 1,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        expect(instance.sortByLessVoted(dumLink1, dumLink2)).toEqual(1);
    });

    it('sortByLessVoted of LinkList while points are equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        instance.sortByUpdatedAt = jest.fn().mockReturnValue(1);
        await expect(instance.sortByLessVoted(dumLink1, dumLink2)).toEqual(1);
        expect(instance.sortByUpdatedAt).toHaveBeenCalledWith(dumLink1, dumLink2);
    });

    it('sortByUpdatedAt of LinkList', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();
        const dumLink1 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        const dumUpdatedDate2 = new Date(dumUpdatedDate.getDate() + 1);

        const dumLink2 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate2,
        };

        await expect(instance.sortByUpdatedAt(dumLink1, dumLink2)).toBeLessThan(0);
    });

    it('sortByCreatedAt of LinkList', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();
        const dumLink1 = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        const dumCreatedDate2 = new Date(dumCreatedDate.getDate() + 1);

        const dumLink2 = {
            createdAt: dumCreatedDate2,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        await expect(instance.sortByCreatedAt(dumLink1, dumLink2)).toBeLessThan(0);
    });

});