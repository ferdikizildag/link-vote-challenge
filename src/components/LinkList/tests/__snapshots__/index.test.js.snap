// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkList renders without crashing 1`] = `
<div>
  <select
    className="orderSelect"
    onChange={[Function]}
  >
    <option
      value=""
    >
      Order by
    </option>
    <option
      value="MOST_VOTED"
    >
      Most Voted (Z-&gt;A)
    </option>
    <option
      value="LESS_VOTED"
    >
      Less Voted (A-&gt;Z)
    </option>
  </select>
  
  <CustomModal
    data={Object {}}
    header="Remove Link"
    onClickCancel={[Function]}
    onClickOk={[Function]}
    question="Do you want to remove:"
    visibility={false}
  />
</div>
`;
